src/main/webapp/WEB-INF/view/jsp/default/ui/casLoginView.jsp

```diff
@@ -50,68 +52,124 @@
+							</div>
+							<input type="hidden" name="lt" value="${loginTicket}"/>
+							<input type="hidden" name="execution" value="${flowExecutionKey}"/>
+							<input type="hidden" name="_eventId" value="submit"/>
+						</form:form>
 					</div>
-					<input type="hidden" name="lt" value="${loginTicket}"/>
-					<input type="hidden" name="execution" value="${flowExecutionKey}"/>
-					<input type="hidden" name="_eventId" value="submit"/>
-				</form:form>
-			</div>
-		</div>
-		<c:if test="${casProperties.OAuthEnabled}">
-			<div class="row">
-				<div class="col-sm-6 col-sm-offset-3 text-center">
-					<h4>or sign in with</h4>
-					<c:if test="${casProperties.googleAuthEnabled}">
-						<a href="${Google2ProviderUrl}" class="btn btn-social-icon btn-lg btn-google-plus" title="Sign in with Google">
-							<i class="fa fa-google-plus"></i>
-						</a>
-					</c:if>
-					<c:if test="${casProperties.gitHubAuthEnabled}">
-						<a href="${GitHubProviderUrl}" class="btn btn-social-icon btn-lg btn-github" title="Sign in with GitHub">
-							<i class="fa fa-github"></i>
-						</a>
-					</c:if>
-					<c:if test="${casProperties.twitterAuthEnabled}">
-						<a href="${SSLTwitterProviderUrl}" class="btn btn-social-icon btn-lg btn-twitter" title="Sign in with Twitter">
-							<i class="fa fa-twitter"></i>
-						</a>
-					</c:if>
-					<c:if test="${casProperties.globusAuthEnabled}">
-						<a href="${GlobusProviderUrl}"><!--<img src="${pageContext.request.contextPath}/images/globus_45.png" width="45" height="45" title="Sign in with Globus" alt="Globus">-->Globus</a>
-					</c:if>
 				</div>
-			</div>
-		</c:if>
+				<c:if test="${casProperties.OAuthEnabled}">
+					<div class="row">
+						<div class="col-sm-6 col-sm-offset-3 text-center">
+							<h4>or sign in with</h4>
+							<c:if test="${casProperties.googleAuthEnabled}">
+								<a href="${Google2ProviderUrl}" class="btn btn-social btn-lg btn-google-plus" title="Sign in with Google">
+									<i class="fa fa-google-plus"></i>
+								</a>
+							</c:if>
+							<c:if test="${casProperties.gitHubAuthEnabled}">
+								<a href="${GitHubProviderUrl}" class="btn btn-social btn-lg btn-github" title="Sign in with GitHub">
+									<i class="fa fa-github"></i>
+								</a>
+							</c:if>
+							<c:if test="${casProperties.twitterAuthEnabled}">
+								<a href="${SSLTwitterProviderUrl}" class="btn btn-social btn-lg btn-twitter" title="Sign in with Twitter">
+									<i class="fa fa-twitter"></i>
+								</a>
+							</c:if>
+							<c:if test="${casProperties.globusAuthEnabled}">
+								<a href="${GlobusProviderUrl}"><!--<img src="${pageContext.request.contextPath}/images/globus_45.png" width="45" height="45" title="Sign in with Globus" alt="Globus">-->Globus</a>
+							</c:if>
+						</div>
+					</div>
+				</c:if>
+			</c:when>
+			<c:otherwise>
+				<c:if test="${casProperties.OAuthEnabled}">
+					<c:choose>
+						<c:when test="${casProperties.OAuthDirectLogin}">
```

Comment:
This logic belongs up top so we can be sure that the redirect occurs before anything gets displayed.